llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -11773,6 +11773,14 @@ void SelectionDAGISel::LowerArguments(const Function &F) {
           AssertOp = ISD::AssertSext;
         else if (Arg.hasAttribute(Attribute::ZExt))
           AssertOp = ISD::AssertZext;
+        if (Arg.hasAttribute(Attribute::NoFPClass)) {
+          SDNodeFlags InValFlags = InVals[i]->getFlags();
+          InValFlags.setNonNeg(Arg.getNoFPClass() & llvm::fcNegative);
+          InValFlags.setNoNaNs(Arg.getNoFPClass() & llvm::fcNan);
+          InValFlags.setNoInfs(Arg.getNoFPClass() & llvm::fcInf);
+          InValFlags.setNoSignedZeros(Arg.getNoFPClass() & llvm::fcNegZero);
```

Comment:
This looks weird. What exactly does the `NoSignedZeros` flag mean on an InVal? Surely it does not just mean "the value is known not to be -0.0"?

Reply:
Yes, "the value is known not to be -0.0" is one case, and another case is that "we don't care about it".
For example `fminimum(3)` requires `-0.0<+0.0`, while in some case, we don't care about it.

Reply:
> Yes, "the value is known not to be -0.0" is one case

I don't think that's true - unless you can point me at some evidence? (I am more familiar with how these flags are defined in IR and MIR than in SDNodes.) I think the NoSignedZeros flag only makes sense on certain _operations_ like FADD, not on arbitrary _values_.

Reply:
Maybe I am wrong.  In https://llvm.org/docs/LangRef.html, there is something about `nofpclass`.
It supports `nzero` and `pzero`. Maybe I misunderstand them.

Reply:
Maybe I need to define another flag for my 2 cases?

Reply:
You can't infer no signed zeros from the fp classes.

I believe the non-neg flag specifically is for integer contexts, and I don't think should be repurposed for FP values this way.

The no-nans/no-infs case is plausible, but this handling is incorrect. You need to ensure that both bits are set (e.gi getNoFPClass & fcInf == fcInf)